src/bedrock/monitor/policy_algorithm.cpp

```diff
@@ -0,0 +1,274 @@
+  logger->info("Promoting {} keys into {} memory slots.", total_rep_to_change, slot);
+  auto time_elapsed = chrono::duration_cast<std::chrono::seconds>(std::chrono::system_clock::now()-grace_start).count();
+
+  if (overflow && adding_memory_node == 0 && time_elapsed > GRACE_PERIOD) {
+    unsigned long long promote_data_size = total_rep_to_change * VALUE_SIZE;
+    unsigned total_memory_node_needed = ceil((ss.total_memory_consumption + promote_data_size) / (MEM_CAPACITY_MAX * tier_data_map[1].node_capacity_));
+
+    if (total_memory_node_needed > memory_node_number) {
+      unsigned node_to_add = (total_memory_node_needed - memory_node_number);
+      logger->info("Insufficient memory nodes: Adding {} memory nodes.", to_string(node_to_add));
+
+      string shell_command = "curl -X POST http://" + management_address + "/add/memory/" + to_string(node_to_add) + " &";
+      system(shell_command.c_str());
+      adding_memory_node = node_to_add;
+    }
+  }
+
+  requests.clear();
+  total_rep_to_change = 0;
+
+  // 3. check key access summary to demote cold keys to ebs tier
+  time_elapsed = chrono::duration_cast<std::chrono::seconds>(std::chrono::system_clock::now()-grace_start).count();
+
+  if (time_elapsed > GRACE_PERIOD) {
+    unsigned slot = (EBS_CAPACITY_MAX * tier_data_map[2].node_capacity_ * ebs_node_number - ss.total_ebs_consumption) / VALUE_SIZE;
+    bool overflow = false;
+
+    for (auto it = key_access_summary.begin(); it != key_access_summary.end(); it++) {
+      string key = it->first;
+      unsigned total_access = it->second;
+
+      if (!is_metadata(key) && total_access < DEMOTE_THRESHOLD && placement[key].global_replication_map_[1] > 0) {
+        total_rep_to_change += 1;
+
+        if (total_rep_to_change > slot) {
+          overflow = true;
+        } else {
+          requests[key] = create_new_replication_vector(0, MINIMUM_REPLICA_NUMBER, 1, 1);
+        }
+      }
+    }
+
+    change_replication_factor(requests, global_hash_ring_map, local_hash_ring_map, routing_address, placement, pushers, mt, response_puller, logger, rid);
+    logger->info("Demoting {} keys into {} EBS slots.", total_rep_to_change, slot);
+    if (overflow && adding_ebs_node == 0) {
+      unsigned long long demote_data_size = total_rep_to_change * VALUE_SIZE;
+      unsigned total_ebs_node_needed = ceil((ss.total_ebs_consumption + demote_data_size) / (EBS_CAPACITY_MAX * tier_data_map[2].node_capacity_));
+
+      if (total_ebs_node_needed > ebs_node_number) {
+        unsigned node_to_add = (total_ebs_node_needed - ebs_node_number);
+        logger->info("Insufficient EBS nodes; Adding {} EBS nodes.", to_string(node_to_add));
+
+        string shell_command = "curl -X POST http://" + management_address + "/add/ebs/" + to_string(node_to_add) + " &";
+        system(shell_command.c_str());
+        adding_ebs_node = node_to_add;
+      }
+    }
+  }
+
+  requests.clear();
+  total_rep_to_change = 0;
+
+  // 4.1 if latency is too high
```

Comment:
Comment style nit: Why do we go from 1, 2, 3 to 4.1, 4.2, etc?

Reply:
fixed